providers/Spotify/mod.ts

```diff
@@ -0,0 +1,321 @@
+					this.lookup.value = this.lookup.value.padStart(14, '0');
+				} else {
+					// No results found
+					break;
+				}
+			}
+
+			// No results found
+			if (!albumId) {
+				throw new ResponseError(this.provider.name, 'API returned no results', this.constructReleaseApiUrl());
+			}
+
+			// Result is a SimplifiedAlbum. Perform a regular ID lookup with the found release
+			// ID to retrieve complete data.
+			this.lookup.method = 'id';
+			this.lookup.value = albumId;
+		}
+
+		const cacheEntry = await this.provider.query<Album>(
+			this.constructReleaseApiUrl(),
+			this.options.snapshotMaxTimestamp,
+		);
+		const release = cacheEntry.content;
+
+		this.updateCacheTime(cacheEntry.timestamp);
+		return release;
+	}
+
+	private async getRawTracklist(rawRelease: Album): Promise<Track[]> {
+		const allTracks: SimplifiedTrack[] = [...rawRelease.tracks.items];
+
+		// The initial response contains max. 50 tracks. Fetch the remaining
+		// tracks with separate requests if needed.
+		let nextUrl = rawRelease.tracks.next;
+		while (nextUrl && allTracks.length < rawRelease.tracks.total) {
+			const cacheEntry = await this.provider.query<ResultList<SimplifiedTrack>>(
+				new URL(nextUrl),
+				this.options.snapshotMaxTimestamp,
+			);
+			this.updateCacheTime(cacheEntry.timestamp);
+			allTracks.push(...cacheEntry.content.items);
+			nextUrl = cacheEntry.content.next;
+		}
+
+		// Load full details including ISRCs
+		if (this.options.withISRC) {
+			return this.getRawTrackDetails(allTracks);
+		} else {
+			return allTracks.map((track) => track as Track);
+		}
+	}
+
+	private async getRawTrackDetails(simplifiedTracks: SimplifiedTrack[]): Promise<Track[]> {
+		const allTracks: Track[] = [];
+		const trackIds = simplifiedTracks.map((track) => track.id);
+
+		// The SimplifiedTrack entries do not contain ISRCs.
+		// Perform track queries to obtain the full details of all tracks.
+		// Each query can return up to 50 tracks.
+		const maxResults = 50;
+		const apiUrl = new URL('tracks', this.provider.apiBaseUrl);
+		for (let index = 0; index < trackIds.length; index += maxResults) {
+			apiUrl.searchParams.set('ids', trackIds.slice(index, index + maxResults).join(','));
+			apiUrl.search = apiUrl.searchParams.toString();
+			const cacheEntry = await this.provider.query<TrackList>(
+				apiUrl,
+				this.options.snapshotMaxTimestamp,
+			);
+			this.updateCacheTime(cacheEntry.timestamp);
+			allTracks.push(...cacheEntry.content.tracks);
+		}
+
+		return allTracks;
+	}
+
+	protected async convertRawRelease(rawRelease: Album): Promise<HarmonyRelease> {
+		this.id = rawRelease.id;
+		const rawTracklist = await this.getRawTracklist(rawRelease);
+		const media = this.convertRawTracklist(rawTracklist);
+		const artwork = selectLargestImage(rawRelease.images, ['front']);
+		return {
+			title: rawRelease.name,
+			artists: rawRelease.artists.map(this.convertRawArtist.bind(this)),
+			gtin: rawRelease.external_ids.ean || rawRelease.external_ids.upc,
```

Comment:
We will loose the `available_markets` then. If you specify the `market` parameter it will only check if the release is available in this market, the result does not contain the `available_markets` lust. If you want to get the list of available markets you must not specify the `market` parameter.

> If a valid user access token is specified in the request header, the country associated with the user account will take priority over this parameter.

We are not using user access tokens. That would require the OAuth flow per user.

Reply:
I think I found a solution to satisfy both. In case of GTIN lookup the provider performs a album search followed by an album ID lookup. We only need the available_markets for the latter case. But the first case is what can sometimes not find a release by UPC.

With the latest change the region is only used for the search request. The lookup by ID still is done without region. In my tests this gives GTIN lookup results for cases a-tisket currently misses. E.g. searching for `810121774182` should find https://open.spotify.com/intl-de/album/2Usbb1Ao4tpzDY0IbIQHqB . This is not available in the US.

A search via `/v1/search?type=album&q=upc%3A0727361584962` gives no result. But including the `market=...` does. It does not even matter what region, both US and non-US does not seem to make a difference.

Once found the request to `/v1/albums/2Usbb1Ao4tpzDY0IbIQHqB` without region works, and includes `available_markets`.

a-tisket currently fails to find a result: https://atisket.pulsewidth.org.uk/?upc=0727361584962&no_deez=no&no_itu=no

Reply:
Sorry, my comment was not very clear, it was a last minute edit before I hit submit and had to run :sweat_smile:
What I meant to say was to only use the `market` parameter for the barcode search, just like you did implement it in the end.

> We are not using user access tokens. That would require the OAuth flow per user.

Thanks for the hint, so I was just confusing two types of tokens here. I interpreted the mentioned "user account" as the one which was used to generate the Spotify developer application, i.e. my own. Now it makes more sense why the search API gave different results than the website search with my account.

Reply:
Final change I hope with 3b2e188ca20dd7606cda08bf843bd3a638a5a1bb. It now searches all given regions (and intentionally no region should no region be given, as this is a valid option for Spotify).

The code does not use the `queryAllRegions` helper because of the interaction with the barcode length variations. It is trying the first region first, there it tries the different barcode lengths. I assume the barcode string mismatches are more common then region errors, and I wanted to avoid querying all regions first with the same wrong barcode before trying the next one.

Reply:
No result: https://atisket.pulsewidth.org.uk/?preferred_countries=US%2CDE&upc=0727361584962&no_deez=no&no_itu=no

Result: http://localhost:8000/release?url=&gtin=0727361584962&region=US%2CDE&spotify=

Reply:
> Thanks for the hint, so I was just confusing two types of tokens here. I interpreted the mentioned "user account" as the one which was used to generate the Spotify developer application, i.e. my own. Now it makes more sense why the search API gave different results than the website search with my account.

Yes, when you mentioned the user tokens the results you get on https://developer.spotify.com/documentation/web-api suddenly also made more sense to me. But I still don't fully understand all the differences. It seems like for searches the API might actually default to US when using client credentials, but not 100% sure. Anyway, the implemented procedure seems to work.